### DISCUSSION TITLE
How to restrict onboarding with external identity providers?

### DISCUSSION THREAD

**Discussion by zookatron** [zitadel/zitadel#6871] [Q&A] [answered]
We need to only allow specific users to create accounts on our Zitadel instance, we can't allow open registration by anybody. This works great when password authentication is used because you can manually create a user in the Zitadel database and they get an email to set up their password. However it seems to be impossible to support this when using external identity providers for login, either we have to allow open registration so that anyone with, for example, a Gmail account can create an account in our system, or the user HAS to first create a password before they can link their Gmail account. Is there any solution for restricted onboarding while using only an external identity provider for the authentication process?

**Comment by hifabienne** [comment]
Hei @zookatron 
I think you should be able to disable the general registration in th elogin policy and only disable it on the idps.
General Setting in Login Policy:
<img width="466" alt="image" src="https://github.com/zitadel/zitadel/assets/38692350/6a34b996-a109-4e61-b801-0c9274a970bb">


IDP Setting:
<img width="466" alt="image" src="https://github.com/zitadel/zitadel/assets/38692350/0ceabf48-a8e9-4b24-8389-8a82b5547266">

**Reply by zookatron** [reply]
Thanks @hifabienne ! Unfortunately if we enable "Account creation allowed" on the IDP, then anyone with an IDP account will be able to create a Zitadel user account, which is not something we can allow. For example, if the IDP is Google, then anyone with a Gmail account can register an account. We need to only allow specific users to have accounts in our system. Domain restriction is not sufficient, we need individual user-level restriction.

**Reply by hifabienne** [reply, marked as answer]
In that case I recommend adding a ZITADEL action, where you can do a check and based on that check return an error or allow it. 
ZITADEL Actions are javascript snippets you can run at certain points within ZITADEL.
https://zitadel.com/docs/apis/actions/introduction

We also have a reporsitory with example actions:
https://github.com/zitadel/actions

**Reply by zookatron** [reply]
Got it, I will look into that, thank you @hifabienne !

**Reply by aharter** [reply]
Hi @hifabienne,
I have a similar use case and was checking out ZITADEL actions. I am not sure how to "return an error" and make it user friendly

I setup a function throwing an error like this.
```
function errorTest(ctx, api) {
  throw new Error("This is an error");
}
```
Setting up as flow "Internal Authentication" with trigger "Pre Creation" works fine. However, the error is shown as `Error: This is an error at errorTest (:3:9(3))`. Is there a way for an action to display an error to the user e.g. returning/throwing a ZITADEL specific error object?

**Reply by hifabienne** [reply]
@livio-a @adlerhurst can you help here?

**Reply by adlerhurst** [reply]
This is not possible at the moment
